### DISCUSSION TITLE
Varifier, primitives, and narrowing conversions

### DISCUSSION THREAD

**Discussion by ben-manes** [google/error-prone#4577] [General]
I assume this has been asked internally but I didn't see an external discussion. Perhaps I am naively or too opinionated, but it feels overly broad that primitive casts are included for `var` assignments. These are usually narrowing conversions such as

```java
int index = (int) (ticks & (wheel[i].length - 1));
```
where ticks is the current nano time and the mask is a fast modulus for indexing into a timing wheel.

There are of course cases where this suggestion makes it longer due to generics, but those seem more reasonable to not over optimize for, is consistent, and the LHS's type length shouldn't be a factor.
```java
V value = (V) ...
```

As such I may be over thinking it, it just feels a little awkward for primitive assignments and even worse to add suppressions lines for each of those.

**Comment by cushon** [comment]
I don't remember seeing discussion about this internally. Some consideration was given to the initial set of heuristics, but I also don't think we've migrated a large amount of code using varifier so far, so I think there's certainly room to change the heuristics.

To check my understanding, varifier has a heuristic to suggest rewriting `Foo foo = (Foo) ...` as `var foo = (Foo) ...` [here](https://github.com/google/error-prone/blob/1ff295936ad30dd83788b5f21ea3f265c3a31f66/core/src/main/java/com/google/errorprone/bugpatterns/Varifier.java#L91-L95), and the feedback is that you'd prefer to avoid the rewrite for primitive types?

I don't think `var foo = (long) ...` or `var foo = (int) ...` is a big readability improvement, but I guess I also see the argument that the type is explicitly provided in the statement and repeating it doesn't necessarily improve readability. Does that capture it, or are there additional reasons why that rewrite should be discouraged?

**Reply by ben-manes** [reply]
It is not a strong opinion towards whether to declare primitives using `var`. It is merely that a narrowing the numeric data type seemed like a less obvious case where `var` provided any readability benefit, given the intent is to avoid boilerplate and that usages are often for bit twiddling. In those cases where the code is low-level, I tend to prefer being very explicit rather than mix-and-match with high-level idioms. Hence I decided to suppress it method-level (whereas typically I try to narrow to the individual line), and all is fine. I am merely trying to default all checks to be enabled and reduce how many were disabled, often because they were non-critical so I didn't want to bother with it, and this case felt a little excessive given _my_ use-case, but that is also probably too opinionated.
